### PR TITLE
fix(compiler): fix determination of when to `super()` in convert-deco…

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
…rators.ts

This fixes an issue where we were inadvertently adding a `super()` call to 'synthetic' constructors created in the `updateConstructor` function.

In particular, this could happen when a class declaration had an `implements` keyword on it, like

```ts
class Foo implements Bar {
    foo = "bar";
}
```

In that case we _don't_ want to add a `super()` call to any constructor we create, because the class has no superclass. So we need to tell that situation apart from

```ts
class Foo extends Bar {
    foo = "bar";
}
```

in which we _do_ want to add a `super()` call.

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Unit tests (`npm test`) were run locally and passed
- [x] E2E Tests (`npm run test.karma.prod`) were run locally and passed
- [x] Prettier (`npm run prettier`) was run locally and passed

## Pull request type

 

 

Please check the type of change your PR introduces:
- [x] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

GitHub Issue Number: N/A

## What is the new behavior?

-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/compiler/transformers/decorators-to-static/convert-decorators.ts [hunk 1]
```diff
@@ -414,8 +414,20 @@ export const updateConstructor = (
  * @param classDeclaration a class declaration AST node
  * @returns whether this class has parents or not
  */
-const needsSuper = (classDeclaration: ts.ClassDeclaration): boolean =>
-  classDeclaration.heritageClauses && classDeclaration.heritageClauses.length > 0;
+const needsSuper = (classDeclaration: ts.ClassDeclaration): boolean => {
+  let hasHeritageClauses = classDeclaration.heritageClauses && classDeclaration.heritageClauses.length > 0;
```
Comment (rwaskiewicz) [inline src/compiler/transformers/decorators-to-static/convert-decorators.ts:418]:
```suggestion
  const hasHeritageClauses = classDeclaration.heritageClauses?.length > 0;
```

File: src/compiler/transformers/decorators-to-static/convert-decorators.ts [hunk 2]
```diff
@@ -414,8 +414,20 @@ export const updateConstructor = (
...
+
+  if (hasHeritageClauses) {
+    // A {@link ts.SyntaxKind.HeritageClause} node may be for extend a
+    // superclass _or_ for implementing an interface. We only want to add a
+    // `super()` call to our synthetic constructor here in the case that there
+    // is a superclass, so we can check for that situation by checking for the
+    // present of a heritage clause with the `.token` property set to
+    // `ts.SyntaxKind.ExtendsKeyword`.
```
Comment (rwaskiewicz) [inline src/compiler/transformers/decorators-to-static/convert-decorators.ts:426]: TIL

**Review by rwaskiewicz — CHANGES_REQUESTED** [review]

File: src/compiler/transformers/decorators-to-static/convert-decorators.ts (hunk 2 above)
Comment (alicewriteswrongs) [reply src/compiler/transformers/decorators-to-static/convert-decorators.ts:426]: me too 🤦‍♀️

Comment (tanner-reits) [reply src/compiler/transformers/decorators-to-static/convert-decorators.ts]: Kinda dumb imo. They should offer a better way of checking this 🙄

**Review by tanner-reits — APPROVED** [review]

**Review by rwaskiewicz — APPROVED** [review]

**Pull request merged**